cmd/zfs/zfs_main.c

```diff
@@ -4383,6 +4383,76 @@ zfs_do_snapshot(int argc, char **argv)
 	return (-1);
 }
 
+typedef struct zfs_send_exclude_arg {
+	size_t count;
+	char **list;
+} zfs_send_exclude_arg_t;
+
+/*
+ * This function creates the zfs_send_exclude_arg_t
+ * object described above; it can be called multiple
+ * times, and the input can be comma-separated.
+ * This is NOT the most efficient data layout; however,
+ * I couldn't think of a non-pathological case where
+ * it should have more than a couple dozen instances
+ * of excludes. If that turns out to be used in
+ * practice, we might want to instead use a tree.
+ */
+static void
+add_dataset_excludes(char *exclude, zfs_send_exclude_arg_t *context)
+{
+	char *tok;
+	while ((tok = strsep(&exclude, ",")) != NULL) {
+		context->list = safe_realloc(context->list,
+		    (sizeof (char *)) * (context->count + 1));
+		context->list[context->count++] = tok;
+	}
+}
+
+static void
+free_dataset_excludes(zfs_send_exclude_arg_t *exclude_list)
+{
+	free(exclude_list->list);
+}
+
+/*
+ * This is the call back used by zfs_send to
+ * determine if a dataset should be skipped.
+ * As stated above, this is not the most efficient
+ * data structure to use, but as long as the
+ * number of excluded datasets is relatively
+ * small (a couple of dozen or so), it won't
+ * have a big impact on performance on modern
+ * processors. Since it's excluding hierarchies,
+ * we'd probably want to move to a more complex
+ * tree structure in that case.
+ */
+static boolean_t
+zfs_do_send_exclude(zfs_handle_t *zhp, void *context)
+{
+	zfs_send_exclude_arg_t *exclude = context;
+	const char *fullname = zfs_get_name(zhp);
+	size_t indx;
+	const char *name = strchrnul(fullname, '/');
+
+	if (exclude == NULL || *name == '\0')
```

Comment:
This still swallows root dataset exclusions – you probably want to forbid those explicitly.

Reply:
Error out if it doesn't have a "/" in it?

Reply:
Error out in `add_dataset_excludes()` if `!zfs_name_valid(tok, ZFS_TYPE_DATASET) || strchr(tok, '/') == NULL`, yeah (this would require taking full dataset paths, of course, but that's more consistent and farms out the validation to libzfs, which is a net plus).